dipy/denoise/localpca.py

```diff
@@ -0,0 +1,137 @@
+                  PCA. PLoS ONE 8(9): e73021.
+                  https://doi.org/10.1371/journal.pone.0073021
+    """
+    if not arr.ndim == 4:
+        raise ValueError("PCA denoising can only be performed on 4D arrays.",
+                         arr.shape)
+
+    patch_size = 2 * patch_radius + 1
+
+    if patch_size ** 3 < arr.shape[-1]:
+        e_s = "You asked for PCA denoising with a patch_radius of {0} "
+        e_s += "for data with {1} directions. This would result in an "
+        e_s += "ill-conditioned PCA matrix. Please increase the patch_radius."
+        e_s.format(patch_radius, arr.shape[-1])
+        raise ValueError(e_s)
+
+    tau = 2.3 * 2.3 * sigma * sigma
+    if isinstance(sigma, np.ndarray) and sigma.ndim == 3:
+        sigma = (np.ones(arr.shape, dtype=np.float64) *
+                 sigma[..., np.newaxis])
+        tau = (np.ones(arr.shape, dtype=np.float64) * tau[..., np.newaxis])
+    else:
+        sigma = np.ones(arr.shape, dtype=np.float64) * sigma
+        tau = np.ones(arr.shape, dtype=np.float64) * tau
+
+    # loop around and find the 3D patch for each direction at each pixel
+
+    # declare arrays for theta and thetax
+    theta = np.zeros(arr.shape, dtype=np.float64)
+    thetax = np.zeros(arr.shape, dtype=np.float64)
+
+    for k in range(patch_radius, arr.shape[2] - patch_radius):
+        for j in range(patch_radius, arr.shape[1] - patch_radius):
+            for i in range(patch_radius, arr.shape[0] - patch_radius):
+
+                X = np.zeros(
+                    (patch_size * patch_size * patch_size, arr.shape[3]))
+                M = np.zeros(arr.shape[3])
+
+                temp = arr[i - patch_radius: i + patch_radius + 1,
+                           j - patch_radius: j + patch_radius + 1,
+                           k - patch_radius: k + patch_radius + 1, :]
+                X = temp.reshape(
+                    patch_size *
+                    patch_size *
+                    patch_size,
+                    arr.shape[3])
+                # compute the mean and normalize
+                M = np.mean(X, axis=0)
+                X = X - M
+                # PCA. There are two ways to do this:
+                # # 1. Compute the covariance matrix C = X_transpose X
+                # C = np.dot(X.T, X)
+                # C = C / X.shape[0]
+                # # and then decompose it:
+                # d, W = sla.eigh(C)
+
+                # 2. Alternatively, calculate the eigenvalues and
+                #  eigenvectors of the covariance matrix through an SVD:
+                U, S, Vt = svd(X, *svd_args)[:3]
+                # Items in S are the eigenvalues, but in ascending order
+                # We invert the order (=> descending), square and normalize
+                # \lambda_i = s_i^2 / n
+                d = S[::-1] ** 2 / X.shape[0]
+                # Rows of Vt are eigenvectors, but also in ascending eigenvalue
+                # order:
+                W = Vt[::-1].T
```

Comment:
Yes. For a nice explanation on the relationship, see also: https://stats.stackexchange.com/a/134283